lib/usb/usb_process.js

```diff
@@ -1,236 +1,271 @@
-var stream = require('stream')
-  , async = require('async')
-  , protocol = require('usb-daemon-parser')
-  , util = require('util')
-  , EventEmitter = require('events').EventEmitter;
+var stream = require('stream'),
+  async = require('async'),
+  protocol = require('usb-daemon-parser'),
+  util = require('util'),
+  EventEmitter = require('events').EventEmitter
   ;
 
+// The largest packet we can send over USB
 var MAX_PACKET_SIZE = 255;
+// The size of the back pressure buffers
 var MAX_BUFFER_SIZE = 4096;
 
 // A process is running remotely on the Tessel and data is piped in over USB
 function USBProcess(id, daemon) {
   var self = this;
+  // The numerical identifer for this process
   this.id = id;
-  // 4 remote process streams
-  this.control = new stream.Writable;
-  this.control.credit = 0;
-  this.control.backPressure = new Array(0);
-  this.stdin = new stream.Writable;
-  this.stdin.credit = 0;
-  this.stdin.backPressure = new Array(0);
-  this.stdout = new stream.Readable;
-  this.stdout.credit = 0;
-  this.stderr = new stream.Readable;
-  this.stderr.credit = 0;
   // Status of whether this process is killed or active
   this.active = true;
-  // Code to be set upon death
-  this.exitCode;
-
-  var self = this;
-
-  this.stdout._read = function() {};
-  this.stderr._read = function() {};
-
-  // This puts the stream into 'flowing mode' 
-  // so that the stream emits ('end') events without 'data' listeners
-  this.stdout.resume();
-  this.stderr.resume();
-
-  this.stdout.signal = function(signalToSend) {
-    switch(signalToSend) {
-      case 'KILL':
-        self.kill(9);
-        break;
-    }
-  }
-
-  // When we receive a stderr write message, we'll want to push it
-  self.on('write', function(packet) {
-    // Push the suggested data into the suggested stream
-    packet.stream.push(packet.data);
-
-    var ackFunc = (packet.stream == self.stdout ? protocol.stdoutAck : protocol.stderrAck);
-
-    self._readableAckHelper(packet.stream , ackFunc, packet.data.length);
-  });
+  // 4 remote process streams
+  this.control = new RemoteWritableStream(id, daemon, protocol.controlWrite, protocol.controlClose);
+  this.stdin = new RemoteWritableStream(id, daemon, protocol.stdinWrite, protocol.stdinClose);
+  this.stdout = new RemoteReadableStream(id, daemon, protocol.stdoutAck);
+  this.stderr = new RemoteReadableStream(id, daemon, protocol.stderrAck);
 
   // Once the process is killed
   self.once('death', function(data) {
-    self.active = false;
-    // Set the exit code
-    self.exitCode = data.arg;
-    // Close the input streams
+    // Close the input streams because the remote proc can't handle data
     self.control.end();
     self.stdin.end();
-    // Clean up remote resources
-    self.close();
+
+    // Mark this process as inactive
+    self.active = false;
```

Comment:
You should also `closeProcessIfReady` here, as it's possible for stdout and stderr to EOF before the process exits.